Sources/PenguinStructures/FixedSizeArray.swift

```diff
@@ -0,0 +1,269 @@
+                          CustomStringConvertible where Index == Int
+{
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.count == c`, where `c` is the capacity of instances.
+  init<Source: Collection>(_ source: Source) where Source.Element == Element
+  // Note: we don't have a generalization to `Sequence` because we couldn't
+  // find an implementation optimizes nearly as well, and in practice
+  // `Sequence`'s that are not `Collection`s are extremely rare.
+
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  init(_ source: ArrayN<Self>, removingAt targetPosition: Index)
+  
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at `targetPosition`.
+  func inserting(_ newElement: Element, at targetPosition: Index) -> ArrayN<Self>
+}
+
+/// Default implementation of `CustomStringConvertible` conformance.
+public extension FixedSizeArray {
+  var description: String { "\(Array(self))"}
+  
+  @_transparent
+  func withUnsafeBufferPointer<R>(
+      _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafePointer(to: self) { p in
+      try body(
+          UnsafeBufferPointer<Element>(
+              start: UnsafeRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  @_transparent
+  mutating func withUnsafeMutableBufferPointer<R>(
+      _ body: (UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafeMutablePointer(to: &self) { p in
+      try body(
+          UnsafeMutableBufferPointer<Element>(
+              start: UnsafeMutableRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at the start.
+  func prepending(_ newElement: Element) -> ArrayN<Self> {
+    .init(head: newElement, tail: self)
+  }
+}
+
+/// A fixed sized collection of 0 elements.
+public struct Array0<T> : FixedSizeArray {
+  public init() {  }
+  
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.isEmpty`
+  @inline(__always)
+  public init<Source: Collection>(_ source: Source)
+      where Source.Element == Element
+  {
+    precondition(source.isEmpty, "Too many elements in source")
+  }
+  
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  public init(_ source: ArrayN<Self>, removingAt targetPosition: Index) {
+    precondition(targetPosition == 0, "Index out of range.")
```

Comment:
```suggestion
    precondition(targetPosition == 0, "Index out of range (\(targetPosition) passed the end).")
```

Reply:
If you mean s/passed/past/ then that's pretty clever!  But also, https://github.com/saeta/penguin/pull/24#discussion_r428787052

Reply:
Indeed I do mean that... 🤦 I do think that putting a little effort into these error messages could be helpful, especially as this should be on the `_slowPath`, and shouldn't slow down typical execution, right?

Reply:
Yes, if carrying information along to produce the error is not an added burden to optimization, including subtle issues like register pressure in a low-level component like this one.  That's why I think what you did is clever.  That said, even what you suggest sacrifices usability a bit since the most useful result tells us what the original index was and how long the array is, and doing *that* without optimization cost is more challenging.  Can you please open an issue for this investigation?